pkg/api/controllers/catalog/catalog.go

```diff
@@ -21,21 +23,107 @@ type CacheCleaner struct {
 	projectCatalogClient v3.ProjectCatalogInterface
 	clusterCatalogClient v3.ClusterCatalogInterface
+	manager              *manager.Manager
 	debounce             func(func())
 }
 
 func Register(ctx context.Context, context *config.ScaledContext) {
+	managementCtx, err := context.NewManagementContext()
+	if err != nil {
+		panic(err)
+	}
 	cleaner := &CacheCleaner{
 		catalogClient:        context.Management.Catalogs(""),
 		projectCatalogClient: context.Management.ProjectCatalogs(""),
 		clusterCatalogClient: context.Management.ClusterCatalogs(""),
+		manager:              manager.New(managementCtx),
 		debounce:             debounce.New(time.Minute),
 	}
 	go cleaner.runPeriodicCatalogCacheCleaner(ctx, time.Hour)
 
 	context.Management.Catalogs("").Controller().AddHandler(ctx, "catalogCache", cleaner.destroyCatalogSync)
 	context.Management.ClusterCatalogs("").Controller().AddHandler(ctx, "clusterCatalogCache", cleaner.destroyClusterCatalogSync)
 	context.Management.ProjectCatalogs("").Controller().AddHandler(ctx, "projectCatalogCache", cleaner.destroyProjectCatalogSync)
+
+	// Don't sync catalog templates on leader node of duplicate actions
+	if context.PeerManager != nil && !context.PeerManager.IsLeader() {
+		logrus.Debug("registering scaled context catalog handler")
+		context.Management.Catalogs("").Controller().AddHandler(ctx, "scaledCatalogSync", cleaner.scaledCatalogSync)
+		context.Management.ClusterCatalogs("").Controller().AddHandler(ctx, "scaledClusterCatalogSync", cleaner.scaledClusterCatalogSync)
+		context.Management.ProjectCatalogs("").Controller().AddHandler(ctx, "scaledProjectCatalogSync", cleaner.scaledProjectCatalogSync)
+	}
+}
+
+func (c *CacheCleaner) scaledCatalogSync(key string, obj *v3.Catalog) (runtime.Object, error) {
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	if obj == nil {
+		return nil, nil
+	}
+
+	// always get a refresh catalog from etcd
+	catalog, err := c.catalogClient.GetNamespaced(ns, name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	commit, helm, err := helmlib.NewForceUpdate(catalog)
+	logrus.Debugf("Helm update local path: %s to commit %s", helm.LocalPath, commit)
+	if err != nil {
+		return c.manager.UpdateCatalogError(catalog, err)
+	}
+	return nil, nil
+}
+
+func (c *CacheCleaner) scaledProjectCatalogSync(key string, obj *v3.ProjectCatalog) (runtime.Object, error) {
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	if obj == nil {
+		return nil, nil
+	}
+
+	// always get a refresh catalog from etcd
+	clusterCatalog, err := c.clusterCatalogClient.GetNamespaced(ns, name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	commit, helm, err := helmlib.NewForceUpdate(&clusterCatalog.Catalog)
+	logrus.Debugf("Helm update local path: %s to commit %s", helm.LocalPath, commit)
+	if err != nil {
+		return c.manager.UpdateClusterCatalogError(clusterCatalog, err)
```

Comment:
Similar to above comment about calling `c.projectCatalogClient`,  `UpdateProjectCatalogError` should be called here and the other way round in `scaledClusterCatalogSync`, I missed that earlier.